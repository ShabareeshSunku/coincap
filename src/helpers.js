function abbrNumber(actual, fixed) {
  const num = parseFloat(actual);
  if (num === null) {
    return null;
  } // terminate early
  if (num === 0) {
    return "0";
  } // terminate early
  fixed = !fixed || fixed < 0 ? 0 : fixed; // number of decimal places to show
  var b = num.toPrecision(2).split("e"), // get power
    k = b.length === 1 ? 0 : Math.floor(Math.min(b[1].slice(1), 14) / 3), // floor at decimals, ceiling at trillions
    c =
      k < 1
        ? num.toFixed(0 + fixed)
        : (num / Math.pow(10, k * 3)).toFixed(1 + fixed), // divide by power
    d = c < 0 ? c : Math.abs(c), // enforce -0 is 0
    e = d + ["", "K", "M", "B", "T"][k]; // append power
  return e;
}
/**
 * Function to process Assets
 * @param {Array} assets assets from the coincap API
 * @returns {Array} properly formatted asset data pushed into an array
 */
function processAssets(assets = []) {
  //variable to hold the formatted(processed) data
  let formattedAssets = [];
  let assetLen = assets.length;
  for (let i = 0; i < assetLen; i++) {
    let ithAsset = assets[i];
    let symbol = ithAsset.symbol.toLowerCase()
    formattedAssets.push({
      assetId: ithAsset.id,
      name: ithAsset.name,
      rank: ithAsset.rank,
      symbol: symbol,
      supply: abbrNumber(ithAsset.supply, 2),
      marketCap: abbrNumber(ithAsset.marketCapUsd, 2),
      volume: abbrNumber(ithAsset.volumeUsd24Hr, 2),
      price: parseFloat(ithAsset.priceUsd).toFixed(2),
      vwap24Hr: abbrNumber(ithAsset.vwap24Hr, 2),
      change: parseFloat(ithAsset.changePercent24Hr).toFixed(2),
      imageUrl: `https://static.coincap.io/assets/icons/${symbol}@2x.png`
    });
  }
  return {
    assets: formattedAssets
  };
}

function processExchanges(exchanges = []) {
  let formattedExchanges = [];
  let exchangeLen = exchanges.length;
  for (let i = 0; i < exchangeLen; i++) {
    let ithExchange = exchanges[i];
    formattedExchanges.push({
      name: ithExchange.name,
      rank: ithExchange.rank,
      tradingPairs: ithExchange.tradingPairs,
      volume: abbrNumber(ithExchange.volumeUsd, 2),
      totalPercent: parseFloat(ithExchange.percentTotalVolume).toFixed(2)
    });
  }
  return {
    exchanges: formattedExchanges
  };
}
export { processAssets, processExchanges };
