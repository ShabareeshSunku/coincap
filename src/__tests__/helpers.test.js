import { processCoinList, processCoin, processExchangesList, processCoinHistory, abbrNumber } from '../helpers'

const coinListMock = [{ "id": "bitcoin", "rank": "1", "symbol": "BTC", "name": "Bitcoin", "supply": "17833575.0000000000000000", "maxSupply": "21000000.0000000000000000", "marketCapUsd": "178726775535.4699808212764900", "volumeUsd24Hr": "6576536582.9421380379100058", "priceUsd": "10021.9263684073429372", "changePercent24Hr": "-2.3446939690666684", "vwap24Hr": "10135.9248452182911382" }, { "id": "ethereum", "rank": "2", "symbol": "ETH", "name": "Ethereum", "supply": "107019652.1240000000000000", "maxSupply": null, "marketCapUsd": "22704921743.3568432983256706", "volumeUsd24Hr": "2142946644.6911915809063679", "priceUsd": "212.1565646377679240", "changePercent24Hr": "-1.4760358265832596", "vwap24Hr": "214.2330964435728628" }, { "id": "ripple", "rank": "3", "symbol": "XRP", "name": "XRP", "supply": "42832704971.0000000000000000", "maxSupply": "100000000000.0000000000000000", "marketCapUsd": "13273148414.3178526807508153", "volumeUsd24Hr": "328601262.6322863506211281", "priceUsd": "0.3098834972786443", "changePercent24Hr": "-2.4818117654853840", "vwap24Hr": "0.3159417798877903" }]
const historyMock = [{ "priceUsd": "9596.1458322208468691", "time": 1563321600000, "date": "2019-07-17T00:00:00.000Z" }, { "priceUsd": "10067.8049747830851174", "time": 1563408000000, "date": "2019-07-18T00:00:00.000Z" }, { "priceUsd": "10487.4592613145577358", "time": 1563494400000, "date": "2019-07-19T00:00:00.000Z" }, { "priceUsd": "10689.0656592893457750", "time": 1563580800000, "date": "2019-07-20T00:00:00.000Z" }, { "priceUsd": "10587.5608352864257851", "time": 1563667200000, "date": "2019-07-21T00:00:00.000Z" }, { "priceUsd": "10468.2066641144659385", "time": 1563753600000, "date": "2019-07-22T00:00:00.000Z" }]
describe('helpers testing', () => {
    it('should return formatted coinList', () => {
        const coins = processCoinList(coinListMock)
        expect(coins).toMatchSnapshot()
    })

    it('should return formatted history', () => {
        const history = processCoinHistory(historyMock)
        expect(history).toMatchSnapshot()
    })

    it('should return abbreviated Number', () => {
        const abbrNo = abbrNumber('1000', 2)
        expect(abbrNo).toBe('1K')
    })
})