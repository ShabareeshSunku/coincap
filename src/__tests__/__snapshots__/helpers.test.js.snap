// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers testing should return formatted coinList 1`] = `
Object {
  "coins": Array [
    Object {
      "assetId": "bitcoin",
      "change": "-2.3446939690666684",
      "imageUrl": "https://static.coincap.io/assets/icons/btc@2x.png",
      "marketCap": "178726775535.4699808212764900",
      "name": "Bitcoin",
      "price": "10021.9263684073429372",
      "rank": "1",
      "supply": "17833575.0000000000000000",
      "symbol": "btc",
      "volume": "6576536582.9421380379100058",
      "vwap24Hr": "10135.9248452182911382",
    },
    Object {
      "assetId": "ethereum",
      "change": "-1.4760358265832596",
      "imageUrl": "https://static.coincap.io/assets/icons/eth@2x.png",
      "marketCap": "22704921743.3568432983256706",
      "name": "Ethereum",
      "price": "212.1565646377679240",
      "rank": "2",
      "supply": "107019652.1240000000000000",
      "symbol": "eth",
      "volume": "2142946644.6911915809063679",
      "vwap24Hr": "214.2330964435728628",
    },
    Object {
      "assetId": "ripple",
      "change": "-2.4818117654853840",
      "imageUrl": "https://static.coincap.io/assets/icons/xrp@2x.png",
      "marketCap": "13273148414.3178526807508153",
      "name": "XRP",
      "price": "0.3098834972786443",
      "rank": "3",
      "supply": "42832704971.0000000000000000",
      "symbol": "xrp",
      "volume": "328601262.6322863506211281",
      "vwap24Hr": "0.3159417798877903",
    },
  ],
}
`;

exports[`helpers testing should return formatted history 1`] = `
Object {
  "data": Array [
    Array [
      1563321600000,
      "9596.1458322208468691",
    ],
    Array [
      1563408000000,
      "10067.8049747830851174",
    ],
    Array [
      1563494400000,
      "10487.4592613145577358",
    ],
    Array [
      1563580800000,
      "10689.0656592893457750",
    ],
    Array [
      1563667200000,
      "10587.5608352864257851",
    ],
    Array [
      1563753600000,
      "10468.2066641144659385",
    ],
  ],
}
`;
